api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+            return contractModel.isAccepted();
+        } else if(proposerVote != null) {
+            //Streamer did not vote. Take Proposer's vote as the final outcome
+            return proposerVote.isViewerFlaggedComplete();
+        } else if(streamerVote != null){
+            //Proposer did not vote. Take Streamer's vote as the final outcome
+            return streamerVote.isViewerFlaggedComplete();
+        }
+        //Default case is to consider the contract completed and release money to the streamer.
+        return true;
+    }
+
+    private boolean proposerMarkedContractCompleted(@NonNull final ContractModel contractModel) {
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getProposer().getId());
+        return proposerVote != null && proposerVote.isViewerFlaggedComplete();
+    }
+
+    private boolean streamerMarkedContractFailed(@NonNull final ContractModel contractModel) {
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getStreamer().getId());
+        return streamerVote != null && !streamerVote.isViewerFlaggedComplete();
+    }
+
+    private boolean userCanVoteOnContract(@NonNull final UserModel voter, @NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+
+        final UUID proposerId = contractModel.getProposer().getId();
+        final boolean isVoterContractProposer = voter.getId().equals(proposerId);
+        final boolean proposerHasVoted = (voteModelRepository.findByContractIdAndVoterId(contractId, proposerId) != null);
+
+        final UUID streamerId = contractModel.getStreamer().getId();
+        final boolean isVoterContractStreamer = voter.getId().equals(streamerId);
+        final boolean streamerHasVoted = (voteModelRepository.findByContractIdAndVoterId(contractId, streamerId) == null);
+
+        final boolean isUserAllowedToVote = isVoterContractProposer || isVoterContractStreamer;
+        if (!isUserAllowedToVote) {
+            log.warn("Unauthorized vote detected. User: {} attempted to vote on contractId: {} without having permissions to do so",
+                    voter.getId(),
+                    contractModel.getId());
+        }
+
+        final boolean isProposerAndHasVoted = isVoterContractProposer && proposerHasVoted;
+        final boolean isStreamerAndHasVoted = isVoterContractStreamer && streamerHasVoted;
+        final boolean userHasVoted = isProposerAndHasVoted || isStreamerAndHasVoted;
+        if(userHasVoted) {
+            log.warn("Multiple votes detected. User: {} attempted to vote on contractId: {} more than once",
+                    voter.getId(),
+                    contractModel.getId());
+        }
+        return isUserAllowedToVote && !userHasVoted;
```

Comment:
I would advocate for moving this logic inside you if statements above, then just returning true here.

Reply:
I liked it for clarity. I think the mthod reads really nice with this line.
```
userCanVoteOnContract(...) {
    ... some logic ...
    return isUserAllowedToVote && !userHasVoted;
}
```
Anyone that looks at this method should hopefully see the answer to if a user can vote or not is based on the user being allowed to vote and not having voted in the past